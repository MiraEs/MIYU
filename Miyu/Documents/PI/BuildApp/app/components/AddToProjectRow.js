import React, {
	Component,
} from 'react';
import PropTypes from 'prop-types';
import {
	StyleSheet,
	TouchableOpacity,
	View,
} from 'react-native';
import {
	ListView,
	Text,
	TextInputWithButton,
} from 'BuildLibrary';
import styles from '../lib/styles';
import helpers from '../lib/helpers';
import Icon from 'react-native-vector-icons/Ionicons';
import TrackingActions from '../lib/analytics/TrackingActions';
import scrollableHelpers from '../lib/ScrollableHelpers';
import { withNavigation } from '@expo/ex-navigation';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { saveShoppingList } from '../actions/ProjectActions';

const componentStyles = StyleSheet.create({
	row: {
		height: 42,
		paddingHorizontal: styles.measurements.gridSpace2,
	},
	rowFlex: {
		flexDirection: 'row',
		alignItems: 'center',
		justifyContent: 'space-between',
	},
	followingTag: {
		marginRight: styles.measurements.gridSpace1,
	},
	emptyView: {
		marginHorizontal: 6,
	},
	groupListView: {
		backgroundColor: styles.colors.greyLight,
		paddingLeft: 17,
	},
	createProjectInputText: {
		height: 42,
		justifyContent: 'center',
		paddingHorizontal: styles.measurements.gridSpace2,
	},
});

export class AddToProjectRow extends Component {

	constructor(props) {
		super(props);
		this.state = {
			groups: new ListView.DataSource({ rowHasChanged: (r1, r2) => r1 !== r2 }),
			data: this.filterEmptyGroups(props.rowData.shoppingLists),
			isCreatingGroup: false,
			projectSelected: !!props.selectedSessionCartIDs,
		};
	}

	componentWillReceiveProps(nextProps) {
		if (this.props.rowData && this.props.rowData.shoppingLists !== nextProps.rowData.shoppingLists) {
			this.setState({ data: this.filterEmptyGroups(nextProps.rowData.shoppingLists)});
		}

		if (this.props.selectedSessionCartIDs !== nextProps.selectedSessionCartIDs) {
			this.setState({ projectSelected: !!nextProps.selectedSessionCartIDs });
		}
	}

	/**
	 * Groups data has an empty group that gets created with the project and should be removed from the data. This unnamed
	 * group is only needed if the user adds an item to the project without selecting a group.
	 */
	filterEmptyGroups = (shoppingLists) => {
		return shoppingLists ? shoppingLists.filter((group) => group.name) : [];
	};

	getGroupData = () => {
		return this.state.groups.cloneWithRows(this.state.data);
	};

	renderFooter = () => {
		return (
			<TextInputWithButton
				ref={(ref) => {
					if (ref) {
						this.addGroupInputRef = ref;
					}
				}}
				onCreate={(name) => {
					this.setState({ isCreatingGroup: true });
					this.props.actions.saveShoppingList({
						name,
						projectId: this.props.rowData.project.id,
					}).then((shoppingList) => {
						this.props.onCreateNewGroupSuccess(shoppingList);
						if (this.addGroupInputRef) {
							this.addGroupInputRef.clearInput();
						}
					}).done(() => this.setState({ isCreatingGroup: false }));
				}}
				buttonText="Create"
				placeholderText="Add a Group"
				analytics={{
					actionName: TrackingActions.ADD_TO_PROJECT_CREATE_GROUP_TAP,
				}}
				onInputFocus={() => {
					if (this.props.listViewRef && this.addGroupInputRef) {
						const scrollResponder = this.props.listViewRef.getScrollResponder();
						scrollableHelpers.scrollRefToKeyboard(scrollResponder, this.addGroupInputRef, {
							offset: 115,
						});
					}
				}}
				isLoading={this.state.isCreatingGroup}
			/>
		);
	};

	renderGroupRow = (rowData) => {
		const selected = this.props.selectedSessionCartIDs && this.props.selectedSessionCartIDs.has(rowData.sessionCart.sessionCartId);
		const color = selected ? 'accent' : 'secondary';
		return (
			<TouchableOpacity onPress={() => this.props.onGroupRowPress(rowData)}>
				<View style={[componentStyles.row, componentStyles.rowFlex]}>
					<Text
						color={color}
						size="small"
					>
						{rowData.name ? rowData.name : 'First Group'}
					</Text>
					{this.renderSelectedCheckMark(selected)}
				</View>
			</TouchableOpacity>
		);
	};

	renderGroupList = () => {
		if (this.state.projectSelected || this.props.expandRow) {
			return (
				<View style={componentStyles.groupListView}>
					<ListView
						dataSource={this.getGroupData()}
						renderRow={this.renderGroupRow}
						renderFooter={this.renderFooter}
						enableEmptySections={true}
						keyboardShouldPersistTaps="handled"
					/>
				</View>
			);
		}
	};

	/**
	 * returning null until the API provides us with information whether is a shared project or not
	 *
	 const isOwner = false;
	 if (isOwner) {
		return (
			<View style={componentStyles.followingTag}>
				<Text
					color="accent"
					size="small"
					fontStyle="italic"
				>
					Following
				</Text>
			</View>
		);
	 }
	 *
	 */
	renderFollowingTag = () => {
		return;
	};

	renderSelectedCheckMark = (selected) => {
		if (selected) {
			return (
				<Icon
					color={styles.colors.accent}
					size={helpers.isIOS() ? 30 : 15}
					name={helpers.getIcon('checkmark')}
				/>
			);
		}
		return (
			<View style={componentStyles.emptyView}/>
		);
	};

	renderProjectName = (rowData) => {
		const color = this.state.projectSelected ? 'accent' : 'secondary';
		return (
			<Text
				color={color}
				size="small"
			>
				{rowData.project.name}
			</Text>
		);
	};

	render() {
		const { rowData, rowID, onRowPress } = this.props;
		return (
			<View>
				<TouchableOpacity
					onPress={() => onRowPress(rowData, rowID)}
				>
					<View style={[componentStyles.row, componentStyles.rowFlex]}>
						{this.renderProjectName(rowData)}
						<View style={componentStyles.rowFlex}>
							{this.renderFollowingTag()}
							{this.renderSelectedCheckMark(this.state.projectSelected)}
						</View>
					</View>
				</TouchableOpacity>
				{this.renderGroupList()}
			</View>
		);
	}
}

AddToProjectRow.propTypes = {
	actions: PropTypes.object,
	expandRow: PropTypes.bool.isRequired,
	listViewRef: PropTypes.object,
	onCreateNewGroupSuccess: PropTypes.func.isRequired,
	onRowPress: PropTypes.func.isRequired,
	onGroupRowPress: PropTypes.func.isRequired,
	rowData: PropTypes.object.isRequired,
	rowID: PropTypes.string.isRequired,
	selectedSessionCartIDs: PropTypes.object,
};

const mapDispatchToProps = (dispatch) => {
	return {
		actions: bindActionCreators({ saveShoppingList }, dispatch),
	};
};

export default withNavigation(connect(null, mapDispatchToProps)(AddToProjectRow));
