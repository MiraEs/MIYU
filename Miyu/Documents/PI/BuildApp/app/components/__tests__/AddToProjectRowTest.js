'use strict';

jest.unmock('../../../app/components/AddToProjectRow');
jest.unmock('react-native');
jest.mock('../../../app/lib/styles');
jest.mock('BuildLibrary');
jest.mock('BuildNative');
jest.mock('../../../app/lib/helpers');
jest.mock('../../../app/services/httpClient', () => ({}));
jest.mock('../../../app/store/configStore', () => ({}));

jest.mock('react-native-vector-icons/Ionicons', () => 'Icon');
import { AddToProjectRow } from '../AddToProjectRow';
import React from 'react';
import renderer from 'react-test-renderer';

const defaultProps = {
	expandRow: false,
	listViewRef: {},
	onRowPress: jest.fn(),
	onCreateNewGroupPress: jest.fn(),
	onCreateNewGroup: jest.fn(),
	onGroupRowPress: jest.fn(),
	sectionID: '',
	rowData: {
		project: {},
		shoppingLists: [],
	},
	rowID: '',
};

describe('AddToProjectRow component', () => {

	it('should render AddToProjectRow with required props', () => {
		const tree = renderer.create(<AddToProjectRow {...defaultProps} />);
		expect(tree).toMatchSnapshot();
	});

	describe('filterEmptyGroups', () => {
		it('should return empty array', () => {
			const tree = renderer.create(<AddToProjectRow {...defaultProps} />);
			const result = tree.getInstance().filterEmptyGroups(null);
			expect(result).toEqual([]);
		});

		it('should return filtered array', () => {
			const tree = renderer.create(<AddToProjectRow {...defaultProps} />);
			const result = tree.getInstance().filterEmptyGroups([
				{
					name: null,
				},
				{
					name: undefined,
				},
				{
					name: '',
				},
				{
					name: 'test group 1',
				},
			]);
			expect(result).toEqual([{ name: 'test group 1' }]);
		});
	});
});
