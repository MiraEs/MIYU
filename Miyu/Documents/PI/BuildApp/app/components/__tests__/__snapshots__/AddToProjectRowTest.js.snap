// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddToProjectRow component should render AddToProjectRow with required props 1`] = `
<View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View
      style={
        Array [
          Object {
            "height": 42,
            "paddingHorizontal": 16,
          },
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          },
        ]
      }
    >
      <Text
        color="secondary"
        size="small"
      />
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <View
          style={
            Object {
              "marginHorizontal": 6,
            }
          }
        />
      </View>
    </View>
  </View>
</View>
`;
